### ISSUE TITLE
Macros not expanded in #pragma when (just) preprocessing

### ISSUE THREAD

**Issue by urnathan** [llvm/llvm-project#65992]
```
#define PP X
void F() {
#pragma unroll PP
  ;
}
```

We get the expected error, when compiling normally:
```
clang++ -c u.cc
u.cc:7:18: error: use of undeclared identifier 'X'
```

But when preprocessing first, or using -save-temps, compilation will fail complaining about `PP`, not `X`
```
clang++ -E u.cc | clang++ -x cpp-output - -c
u.cc:7:16: error: use of undeclared identifier 'PP'
```

```
clang++ -c -save-temps u.cc                 
u.cc:3:16: error: use of undeclared identifier 'PP'
```

(This breaks -save-temps use when one has pragmas that look like that.)

**Comment by cor3ntin** [comment]
Neither GCC nor Clang replace `PP` in `-E` mode.
But ICC does https://compiler-explorer.com/z/hbWodz61q 

It makes me wonder whether we are purposefully imitating GCC here
@AaronBallman

**Comment by AaronBallman** [comment]
Well that's fun. The preprocessor specification for C and C++ have diverged somewhat, with C providing more information in the form of this note:

> 222) An implementation is not required to perform macro replacement in pragmas, but it is permitted except for in standard
pragmas (where STDC immediately follows pragma). If the result of macro replacement in a non-standard pragma has the
same form as a standard pragma, the behavior is still implementation-defined; an implementation is permitted to behave as
if it were the standard pragma, but is not required to.

I wonder how we all do on handling the `STDC` behavior... https://godbolt.org/z/ffKKr4dKM -- we all seem to get that correct. Oh interesting, EDG's behavior seems to be a bit more complicated though: https://godbolt.org/z/4a8Kc3ocr

Also interesting, according to [GCC's documentation](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#index-fdirectives-only):
> With -E, preprocessing is limited to the handling of directives such as #define, #ifdef, and #error. Other preprocessor operations, such as macro expansion and trigraph conversion are not performed. In addition, the -dD option is implicitly enabled. 

which implies that `-E` is expected to do all preprocessing including macro expansion, and you need to use `-fdirectives-only` to suppress the expansion. Clang definitely is not doing that: https://godbolt.org/z/4bz3eTG3x and neither is GCC: https://godbolt.org/z/o6corceob so perhaps that's a bug with GCC's documentation?

**Comment by urnathan** [comment]
I think the C wording you quote is orthogonal to this issue -- namely if the implementation macro expands, it should do so consistently, regardless of whether one's doing integrated preprocessing or doing it as a separate step. The documented GCC behaviour seems unhelpful

**Comment by AaronBallman** [comment]
> I think the C wording you quote is orthogonal to this issue -- namely if the implementation macro expands, it should do so consistently, regardless of whether one's doing integrated preprocessing or doing it as a separate step. 

Agreed, I was pointing it out because I was curious if the standard mandated anything in the area.

>The documented GCC behaviour seems unhelpful

I think the GCC documented behavior of expanding macros unless you pass `-fdirectives-only` is reasonable, but I think the GCC and Clang behavior of being inconsistent with expansions is not helpful. I think `-E` should give you the expanded preprocessed code, so we should be consistently expanding macros in pragmas (except for `STDC` pragmas, where it's prohibited).

**Comment by urnathan** [comment]
I agree, somebody should poke a GCC developer :)

**Comment by cor3ntin** [comment]
In  `clang::DoPrintPreprocessedInput`, which is what  produces `-E` we set up an `UnknownPragmaHandler`
that never does expansion, except in microsoft mode.

```cpp
// Expand macros in pragmas with -fms-extensions.  The assumption is that
// the majority of pragmas in such a file will be Microsoft pragmas.
```

To fix Nathan's bug, we can change that to always do macro expansion in -E (except for STDC).
Thoughts? Are there other pragmas that should never be macro expanded?


> I think the GCC documented behavior of expanding macros unless you pass `-fdirectives-only` is reasonable, but I think the GCC and Clang behavior of being inconsistent with expansions is not helpful. I think `-E` should give you the expanded preprocessed code, so we should be consistently expanding macros in pragmas (except for `STDC` pragmas, where it's prohibited).

The behavior I'm seeing and what clang seems to be doing is that `-fdirectives-only` does full preprocessing (well, except for pragmas) in directives, and 0 preprocessing outside of directives. Which make sense

I think what is meant is

> With -E, preprocessing is limited to the handling of directives such as #define, #ifdef, and #error. Other preprocessor operations, such as macro expansion **outside of directives** and trigraph conversion are not performed. In addition, the -dD option is implicitly enabled.

**Comment by AaronBallman** [comment]
Related:
It looks like we used to lex unexpanded tokens at the beginning of time, https://github.com/llvm/llvm-project/commit/09e3cdf9ef918df66413e634fc296a356fc1c112

Then we changed it in https://github.com/llvm/llvm-project/commit/0e73ec45502af2d7cf4cd459cf3cc70a8509faf3 claiming that GCC never expands tokens but MSVC does.

Also, OpenMP requires expanding macros, https://github.com/llvm/llvm-project/commit/eab747b9f5d4b6f44a4fa0151b3f9339b6cd564a

> The behavior I'm seeing and what clang seems to be doing is that -fdirectives-only does full preprocessing (well, except for pragmas) in directives, and 0 preprocessing outside of directives. Which make sense

Hmmm, I'm not seeing that happening: https://godbolt.org/z/fhPT9W4Kx -- the output for the pragmas is unexpanded in both forms.

**Comment by cor3ntin** [comment]
> Then we changed it in https://github.com/llvm/llvm-project/commit/0e73ec45502af2d7cf4cd459cf3cc70a8509faf3 claiming that GCC never expands tokens but MSVC does.

So i was correct to assume GCC compat here. We should ask GCC what they mean to do @urnathan 

> Hmmm, I'm not seeing that happening: https://godbolt.org/z/fhPT9W4Kx -- the output for the pragmas is unexpanded in both forms.

**Excluding pragmas**  (pragma are differently broken everywhere) https://godbolt.org/z/bPEbo9Y69

**Comment by urnathan** [comment]
it seems gcc only has `#pragma GCC unroll`, which consistently does not expand macros.  `#pragma unroll` is an unknown pragma.
